bpf/protocol_common.h

```diff
@@ -95,114 +97,189 @@ static __always_inline http_connection_metadata_t *connection_meta_by_direction(
+    bool user_backed;
+    union {
+        size_t iov_offset;
+        int last_offset;
+    };
+    size_t count;
+    union {
+        const struct iovec *iov;
+        const struct kvec *kvec;
+        const struct bio_vec *bvec;
+        struct xarray *xarray;
+        struct pipe_inode_info *pipe;
+        void *ubuf;
+    };
+    union {
+        unsigned long nr_segs;
+        struct {
+            unsigned int head;
+            unsigned int start_head;
+        };
+        loff_t xarray_start;
+    };
+};
 
-static __always_inline int read_msghdr_buf(struct msghdr *msg, u8* buf, int max_len) {
-    struct iov_iter msg_iter = BPF_CORE_READ(msg, msg_iter);
-    u8 msg_iter_type = 0;
-
-    if (bpf_core_field_exists(msg_iter.iter_type)) {
-        bpf_probe_read(&msg_iter_type, sizeof(u8), &(msg_iter.iter_type));
-        bpf_dbg_printk("msg iter type exists, read value %d", msg_iter_type);
-    }
-
-    bpf_dbg_printk("iter type %d", msg_iter_type);
 
-    struct iovec *iov = NULL;
+// older struct that features 'type' instead of 'iter_type'
+struct iov_iter___v58 {
+    unsigned int type;
+    size_t iov_offset;
+    size_t count;
+    union {
+        const struct iovec *iov;
+        const struct kvec *kvec;
+        const struct bio_vec *bvec;
+        struct pipe_inode_info *pipe;
+    };
+    union {
+        unsigned long nr_segs;
+        struct {
+            unsigned int head;
+            unsigned int start_head;
+        };
+    };
+};
 
-    u32 l = max_len;
-    bpf_clamp_umax(l, IO_VEC_MAX_LEN);
+// helper struct used by get_iovec_ctx
+struct iovec_iter_ctx {
+    unsigned int iter_type;
+    size_t iov_offset;
+    size_t count;
+    unsigned long nr_segs;
+    const struct iovec *iov;
+    const void *ubuf;
+};
 
-    if (bpf_core_field_exists(msg_iter.iov)) {
-        bpf_probe_read(&iov, sizeof(struct iovec *), &(msg_iter.iov));
-        bpf_dbg_printk("iov exists, read value %llx", iov);
+// extracts kernel specific iov_iter information into a iovec_iter_ctx instance
+static __always_inline void get_iovec_ctx(struct iovec_iter_ctx* ctx, struct msghdr *msg) {
+    if (LINUX_KERNEL_VERSION <= KERNEL_VERSION(5, 13, 0)) {
+        struct iov_iter___v58 iter;
+        bpf_core_read(&iter, sizeof(iter), &msg->msg_iter);
+
+        ctx->iter_type = iter.type;
+        ctx->iov_offset = iter.iov_offset;
+        ctx->count = iter.count;
+        ctx->nr_segs = iter.nr_segs;
+        ctx->iov = iter.iov;
+        ctx->ubuf = NULL;
+    } else if (LINUX_KERNEL_VERSION <= KERNEL_VERSION(6, 3, 0)) {
+        struct iov_iter___v60 iter;
+        bpf_core_read(&iter, sizeof(iter), &msg->msg_iter);
+
+        ctx->iter_type = iter.iter_type & 0xff;
+        ctx->iov_offset = iter.iov_offset;
+        ctx->count = iter.count;
+        ctx->nr_segs = iter.nr_segs;
+        ctx->iov = iter.iov;
+        ctx->ubuf = iter.ubuf;
     } else {
-        // TODO: I wonder if there's a way to check for field existence without having to
```

Comment:
Perhaps the entire struct does not need to be define, just the new member, i.e.

```
struct iov_iter___v64 {
 union { struct iovec __ubuf_iovec; }
}

if (bpf_core_field_exists(((struct iov_iter___v64*)(0))->type) { ... }
```

I know this works for enums, but I haven't tried it for structs.